File: Python_Heat_Transfer/Heat_Transfer.py
# To install packages, run these commands in your terminal, not in a Python script:
import time
t = time.time()
import numpy as np
import matplotlib.pyplot as plt

# Define thermal diffusivity (aluminum)
a = 97  # mm^2/s
alpha = a / 1000**2  # m^2/s

# Define area matrix, length, time parameters
s = 50  # grid size
dx = 0.01
dy = 0.01
Lx = np.arange(0, (s-1)*dx + dx, dx)
Ly = np.arange(0, (s-1)*dy + dy, dy)
X, Y = np.meshgrid(Lx, Ly)
dt = 0.1

# Initial temperature and heat source
T2 = np.ones((s+2, s+2)) * 273.15  # K
T1 = np.zeros((s+2, s+2))
S = np.zeros((s+2, s+2))
S[s//2, s//2] = 100  # heat source at center

# Boundary conditions: left, bottom, right, top
bdry = [500, 300, 200, 700]
T2[:, 0] = bdry[0]
T2[0, :] = bdry[1]
T2[:, -1] = bdry[2]
T2[-1, :] = bdry[3]
k = 0
value = 1
fig, ax = plt.subplots()
cb = None

def plot_heat():
    global cb
    ax.clear()
    cax = ax.contourf(
        Lx, Ly, T2[1:-1, 1:-1], 20, cmap='jet',
        vmin=min(bdry), vmax=max(bdry)
    )
    ax.set_xlabel('X (m)')
    ax.set_ylabel('Y (m)')
    ax.set_title(f"{k*dt:.1f} (Sec)")
    # Remove previous colorbar if it exists
    #if cb is not None:
        #cb.remove()
    #cb = plt.colorbar(cax, ax=ax)
    #cb.set_label('Temperature (K)')


while value > 0.005:
    T1 = T2.copy()
    T2[1:-1, 1:-1] = (
        T1[1:-1, 1:-1]
        + alpha * dt * (
            (T1[2:, 1:-1] - 2 * T1[1:-1, 1:-1] + T1[:-2, 1:-1]) / dx**2
            + (T1[1:-1, 2:] - 2 * T1[1:-1, 1:-1] + T1[1:-1, :-2]) / dy**2
        )
        + S[1:-1, 1:-1] * dt
    )
    k += 1
    T2[:, 0] = bdry[0]
    T2[0, :] = bdry[1]
    T2[:, -1] = bdry[2]
    T2[-1, :] = bdry[3]
    value = np.abs(np.mean(T2 - T1))
    if k % 10 == 0:
        plot_heat()
        plt.pause(0.0000000000000000000000000000001)

# Show the final plot after the loop
plot_heat()
plt.show(block=False)
plt.close()
elapsed = time.time() - t
print(elapsed)

File: Phase_1_to_3_MATLAB/Heat_transfer.m
function [T]=Heat_transfer(bdry1)

%% define simple constants
%define thermal diffusivity (edit based on material)
%aluminum
a=97; %mm^2/s 
alpha=a/1000^2; %m^2/s

%define area matrix, length, time parameters (cm per side)
s=20; %(edit based on area)

%edit based on persision
%delta_x %delta_y (m)
dx=0.01; dy=0.01;
dt=0.1;

%% define initial and redefine Temperature, and heat source term
T2=ones(s+2)*273.15; %K
[P,Q]=size(T2);
%S=rand(s+2)*30;
S=zeros(s+2);%define S(heat source) array
S(12,3)=10;

k=0;

%% boundary conditions
%edges left bottom right top
bdry=bdry1;
T2(:,1)=bdry(1); T2(1,:)=bdry(2); T2(:,end)=bdry(3); T2(end,:)=bdry(4);  

%% Define Loop
while k<1200 %& abs(sum(T1,"all")-sum(T2,'all'))>0.1
    T1=T2; % set the 2 temp plots equal
       i=2:P-1; %sweep x axis
        j=2:Q-1; %sweep y axis
            %Impliment Explicit Euler Scheme
            T2(i,j)=T1(i,j)+alpha*dt*((T1(i+1,j)-2*T1(i,j)+T1(i-1,j))/dx^2+...
                (T1(i,j+1)-2*T1(i,j)+T1(i,j-1))/dy^2)+S(i,j)*dt; 
%increase tick
k=k+1;
% reimplement boundary conditions
T2(:,1)=bdry(1); T2(1,:)=bdry(2); T2(:,end)=bdry(3); T2(end,:)=bdry(4);
end
T=T2;
end